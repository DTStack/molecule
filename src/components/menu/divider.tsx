import { classNames } from 'mo/common/className';
import React from 'react';
import {
    defaultDividerClassName,
    defaultMenuItemClassName,
    disabledClassName,
} from './base';

const Divider = () => {
    return (
        <li
            className={classNames(defaultMenuItemClassName, disabledClassName)}
            role="separator"
        >
            <a className={defaultDividerClassName} />
        </li>
    );
};

export { Divider };
