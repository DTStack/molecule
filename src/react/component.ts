import { cloneDeep } from 'lodash';
import { EventEmitter, GlobalEvent } from 'mo/common/event';

export enum ComponentEvents {
    Update = 'Component.Update',
}

export interface IComponent<S = any> {
    /**
     * Set the Component state
     * @param values The next values of state
     * @param callback calling after setState
     */
    setState(values: S, callback?: (prevState: S, nextState: S) => void): void;
    /**
     * Trigger the Component update event
     * @param nextState
     */
    render(nextState?: S): void;
    /**
     * Listen to the Component state update event
     * @param listener
     */
    onUpdateState(listener: (prevState: S, nextState: S) => void): void;
    /**
     * Remove the Component update event listening, default is remove all,
     * also you can remove one by pass the listener
     */
    removeOnUpdateState(listener?: Function): void;
    /**
     * Force to update the Component
     */
    forceUpdate(): void;
    /**
     * Get the Component state
     */
    getState(): S;
}

export abstract class Component<S = any>
    extends GlobalEvent
    implements IComponent<S>
{
    protected abstract state: S;
    private _event: EventEmitter;

    constructor() {
        super();
        this._event = new EventEmitter();
    }

    /**
     * Set the state values, and notify the view component to re render
     * @param values update target state values
     */
    public setState(
        values: Partial<S>,
        callback?: (prevState: S, nextState: S) => void
    ) {
        const nextState = Object.assign(this.state, values);
        this.render(nextState);
        callback?.(this.state, nextState);
    }

    /**
     * Initiative notify the component to render the view by the state
     * @param nextState
     */
    public render(nextState?: S) {
        this._event.emit(ComponentEvents.Update, this.state, nextState);
    }

    public onUpdateState(listener: (prevState: S, nextState: S) => void) {
        this._event.subscribe(ComponentEvents.Update, listener);
    }

    public removeOnUpdateState(listener?: Function): void {
        this._event.unsubscribe(ComponentEvents.Update, listener);
    }

    public forceUpdate() {
        this.setState(cloneDeep(this.state));
    }

    public getState(): S {
        return this.state;
    }
}
