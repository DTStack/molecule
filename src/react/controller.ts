import { GlobalEvent } from 'mo/common/event';

export abstract class Controller extends GlobalEvent {
    public abstract initView(): void;
}
