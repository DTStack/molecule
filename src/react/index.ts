export * from './component';
export * from './helper';
export * from './connector';
export * from './controller';
