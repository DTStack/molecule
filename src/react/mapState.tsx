import React, { Component } from 'react';
import Logger from 'mo/common/logger';

/**
 * Mapping the state to the component
 * TODO support mapping service method to the component.
 * @param WrappedComponent The component will be wrapped
 * @param state The state you want to injected, notice the state must be an observable object
 * @param subscribes The events of your subscribe, it used to trigger the state re render
 */
export function mapState<S, T>(
    WrappedComponent: React.ComponentType<S>,
    state: S,
    actions?: T
) {
    return class StateProvider extends Component {
        state: { lastUpdated: number };
        constructor(props) {
            super(props);
            this.onChange = this.onChange.bind(this);
            this.state = {
                lastUpdated: Date.now(),
            };
        }

        componentDidMount() {
            // There is no declare state parameter as IObservable type, so must convert to any type.
            if ((state as any).subscribe) {
                (state as any).subscribe(this.onChange);
            } else {
                Logger.error(
                    'The state parameter of mapState must be an observable object.'
                );
            }
        }
        /**
         * TODO: Performance optimize, now whatever any properties changed in target,
         * there always be trigger the onChange event, so need a compare operation.
         * @param nextState changed data
         */
        onChange(nextState: S) {
            Logger.info(nextState, state);
            this.setState({
                lastUpdated: Date.now(),
            });
        }

        render() {
            return (
                <WrappedComponent
                    {...this.state}
                    {...state}
                    {...this.props}
                    {...actions}
                />
            );
        }
    };
}
