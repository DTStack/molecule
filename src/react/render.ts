import * as ReactDOM from 'react-dom';
import { createRoot, Root } from 'react-dom/client';

const reactdom: typeof ReactDOM & {
    createRoot?: typeof createRoot;
    __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED?: {
        usingClientEntryPoint?: boolean;
    };
} = {
    ...ReactDOM,
};

let nextReactDOMClient: typeof createRoot | undefined;

export const renderedSign = Symbol('__molecule__root');

if (ReactDOM.version.startsWith('18')) {
    nextReactDOMClient = reactdom.createRoot;
}

function toggleWarning(skip: boolean) {
    const { __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED } = reactdom;

    if (
        __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED &&
        typeof __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED === 'object'
    ) {
        __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.usingClientEntryPoint =
            skip;
    }
}

export const render = (
    node: JSX.Element,
    container: HTMLElement & {
        [renderedSign]?: Root;
    }
) => {
    if (nextReactDOMClient) {
        toggleWarning(true);
        const root = container[renderedSign] || nextReactDOMClient(container);
        toggleWarning(false);
        root.render(node);

        container[renderedSign] = root;
        return;
    }

    reactdom.render(node, container);
};

export const unmout = (
    container: HTMLElement & {
        [renderedSign]?: Root;
    }
) => {
    if (nextReactDOMClient) {
        container[renderedSign]?.unmount();
        Reflect.deleteProperty(container, renderedSign);
        return true;
    }

    return reactdom.unmountComponentAtNode(container);
};
