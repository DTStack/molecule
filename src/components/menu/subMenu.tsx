import React from 'react';
import { classNames } from 'mo/common/className';
import { TriggerEvent } from 'mo/common/dom';
import { Icon } from 'mo/components/icon';

import { Menu } from './menu';
import { IMenuItemProps } from './menuItem';
import {
    checkClassName,
    disabledClassName,
    defaultMenuItemClassName,
    defaultSubMenuClassName,
    indicatorClassName,
    labelClassName,
    menuContentClassName,
} from './base';

export enum MenuMode {
    Vertical = 'vertical',
    Horizontal = 'horizontal',
}

export function isHorizontal(mode: MenuMode) {
    return mode === MenuMode.Horizontal;
}

export function isVertical(mode: MenuMode) {
    return mode === MenuMode.Horizontal;
}

export interface ISubMenuProps extends Omit<IMenuItemProps, 'id'> {
    /**
     * The event of show subMenu, default value is 'hover'
     */
    trigger?: TriggerEvent;
    icon?: string | JSX.Element;
    data?: ISubMenuProps[];
    mode?: MenuMode;
}

export function SubMenu(props: React.PropsWithChildren<ISubMenuProps>) {
    const {
        className,
        name,
        render,
        data = [],
        mode = MenuMode.Vertical,
        icon,
        disabled = false,
        children,
        onClick,
        title,
        sortIndex,
        ...restProps
    } = props;
    const cNames = classNames(defaultSubMenuClassName, className);
    const isAlignHorizontal = isHorizontal(mode);

    const chevronType = isAlignHorizontal ? 'down' : 'right';
    const subMenuContent =
        data.length > 0 ? (
            <Menu
                className={cNames}
                style={{ opacity: '0', pointerEvents: 'none' }}
                data={data}
                onClick={onClick}
                {...restProps}
            />
        ) : (
            <Menu
                className={cNames}
                style={{ opacity: '0', pointerEvents: 'none' }}
                onClick={onClick}
            >
                {children}
            </Menu>
        );

    return (
        <li
            className={classNames(
                defaultMenuItemClassName,
                disabled ? disabledClassName : null
            )}
            data-submenu
            data-mode={mode}
            data-sort={sortIndex}
            // prevent render JSX title in HTMLElement
            {...(typeof title === 'string' ? { title } : {})}
            {...restProps}
        >
            <div className={menuContentClassName}>
                {typeof icon === 'string' ? (
                    <Icon className={checkClassName} type={icon || ''} />
                ) : (
                    icon
                )}
                <span className={labelClassName}>
                    {render ? render(props) : name}
                </span>
                <Icon
                    className={indicatorClassName}
                    type={`chevron-${chevronType}`}
                />
            </div>
            {subMenuContent}
        </li>
    );
}
