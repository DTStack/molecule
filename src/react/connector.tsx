import 'reflect-metadata';
import { container } from 'tsyringe';
import React, { Component } from 'react';
import Logger from 'mo/common/logger';
import { IComponent } from './component';
import { Controller } from './controller';

export type ServiceObject = {
    [index: string]: IComponent;
};

export type ControllerObject = {
    [index: string]: Controller;
};

export function connect<T = any>(
    Service: IComponent | ServiceObject,
    View: React.ComponentType<any>,
    Controller?: Controller | ControllerObject
): React.ComponentType<T> {
    return class Connector extends Component<T, any> {
        state: { lastUpdated: number };
        private _isMounted = false;
        constructor(props) {
            super(props);
            this.onChange = this.onChange.bind(this);
            this.state = {
                lastUpdated: Date.now(),
            };
        }

        componentDidMount() {
            this._isMounted = true;
            this.handleService((service) => {
                service.onUpdateState(this.onChange);
            });
        }

        componentWillUnmount() {
            this._isMounted = false;
            this.handleService((service) => {
                service.removeOnUpdateState(this.onChange);
            });
        }

        onChange(prevState, nextState) {
            Logger.info(prevState, nextState, (container as any)._registry);
            this.update();
        }

        update = () => {
            if (this._isMounted) {
                this.setState({
                    lastUpdated: Date.now(),
                });
            }
        };

        getServiceState() {
            const target = {};
            this.handleService((service, prop) => {
                if (prop) {
                    Object.assign(target, {
                        [prop]: { ...service.getState() },
                    });
                } else {
                    Object.assign(target, { ...service.getState() });
                }
            });
            return target;
        }

        handleService(callback: (service: IComponent, prop?: string) => void) {
            if (this.isValidService(Service as IComponent)) {
                callback(Service as IComponent);
            } else {
                for (const name in Service) {
                    if (name) {
                        const service: IComponent = Service[name];
                        if (this.isValidService(service)) {
                            callback(service, name);
                        }
                    }
                }
            }
        }

        getControllerProperties() {
            const target = {};
            if (!Controller) {
                return target;
            }

            this.handleController((controller: Controller, prop) => {
                if (prop) {
                    Object.assign(target, {
                        [prop]: this.getObjectPublicProperties(controller),
                    });
                } else {
                    Object.assign(
                        target,
                        this.getObjectPublicProperties(controller)
                    );
                }
            });
            return target;
        }

        handleController(
            callback: (controller: Controller, prop?: string) => void
        ) {
            if (this.isValidController(Controller as Controller)) {
                callback(Controller as Controller);
            } else {
                for (const name in Controller) {
                    if (name) {
                        const controller: Controller = Controller[name];
                        if (this.isValidController(controller)) {
                            callback(controller, name);
                        }
                    }
                }
            }
        }

        render() {
            return (
                <View
                    {...this.state}
                    {...this.props}
                    {...this.getServiceState()}
                    {...this.getControllerProperties()}
                />
            );
        }

        private isValidService(service: IComponent) {
            return typeof service.onUpdateState === 'function';
        }

        private isValidController(controller: Controller) {
            return typeof controller.initView === 'function';
        }

        private getObjectPublicProperties(obj: Controller) {
            const keys = Object.keys(obj);
            const result = {};
            keys.forEach((key) => {
                // Filter the Service, Controller instances and private properties which start with '_'
                // TODO There need a better way to filter the private properties of Object,
                // maybe we can make use of the # identifier in future.
                if (
                    !key.endsWith('Service') &&
                    !key.endsWith('Controller') &&
                    !key.startsWith('_')
                ) {
                    result[key] = obj[key];
                }
            });
            return result;
        }
    };
}
