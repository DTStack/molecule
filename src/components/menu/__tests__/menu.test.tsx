import React, { useRef } from 'react';
import { fireEvent, render, waitFor, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

import { Menu } from '../';
import { Divider } from '../divider';
import { MenuMode } from '../subMenu';

import {
    activeClassName,
    horizontalMenuClassName,
    verticalMenuClassName,
} from '../base';
import { MenuRef } from '../index';

const menuData = [
    {
        id: 'File',
        name: 'File',
        data: [
            {
                id: 'New File',
                name: 'New File',
            },
            {
                id: 'OpenFile',
                name: 'Open',
            },
        ],
    },
    {
        id: 'Edit',
        name: 'Edit',
        data: [
            {
                id: 'Undo',
                name: 'Undo',
            },
            {
                id: 'Redo',
                name: 'Redo',
            },
        ],
    },
    {
        id: 'Selection',
        name: 'Selection',
        data: [
            {
                id: 'SelectAll',
                name: 'Select All',
            },
            {
                id: 'CopyLineUp',
                name: 'Copy Line Up',
            },
        ],
    },
    {
        id: 'View',
        name: 'View',
        data: [
            {
                id: 'Command Palette',
                name: 'Command Palette',
            },
            {
                id: 'OpenView',
                name: 'Open View',
            },
            {
                id: 'Appearance',
                name: 'Appearance',
                data: [
                    {
                        icon: 'check',
                        id: 'ShowMenuBar',
                        name: 'Show Menu Bar',
                    },
                    {
                        icon: 'check',
                        id: 'ShowSideBar',
                        name: 'Show Side Bar',
                    },
                    {
                        icon: 'check',
                        id: 'ShowStatusBar',
                        name: 'Show Status Bar',
                    },
                    {
                        icon: 'check',
                        id: 'ShowActivityBar',
                        name: 'Show Activity Bar',
                    },
                ],
            },
        ],
    },
    {
        id: 'Run',
        name: 'Run',
        data: [
            {
                id: 'RunTask',
                name: 'Run Task',
            },
        ],
    },
    {
        id: 'Help',
        name: 'Help',
        data: [
            {
                id: 'About',
                name: 'About',
            },
        ],
    },
];
const TEST_ID = 'test-id';

function MenuTest(props) {
    const ref = useRef<MenuRef>(null);
    return <Menu ref={ref} data={menuData} {...props} />;
}

describe('Test the Menu Component', () => {
    test('Match the List snapshot', () => {
        const component = renderer.create(
            <Menu style={{ width: 200 }} data={menuData} />
        );
        const menu = component.toJSON();

        expect(menu).toMatchSnapshot();
    });

    test('Match the divider snapshot', () => {
        const component = renderer.create(<Divider />);
        const divider = component.toJSON();

        expect(divider).toMatchSnapshot();
    });

    test('Set className to Menu', () => {
        const classname = 'menu';
        const wrapper = render(
            <Menu data-testid={TEST_ID} className={classname} />
        );
        const component = wrapper.getByTestId(TEST_ID);

        expect(component.classList).toContain(classname);
    });

    test('Set horizontal to Menu', () => {
        const mode = MenuMode.Horizontal;
        const wrapper = render(<Menu data-testid={TEST_ID} mode={mode} />);
        const component = wrapper.getByTestId(TEST_ID);

        expect(component.classList).toContain(horizontalMenuClassName);
    });

    test('Set vertical to Menu', () => {
        const mode = MenuMode.Vertical;
        const wrapper = render(<Menu data-testid={TEST_ID} mode={mode} />);
        const component = wrapper.getByTestId(TEST_ID);

        expect(component.classList).toContain(verticalMenuClassName);
    });

    test('Set data single to Menu', () => {
        const mockData = [{ id: 'Edit', name: 'Edit', title: TEST_ID }];
        const wrapper = render(<Menu data={mockData} />);
        const component = wrapper.getByTitle(TEST_ID);

        expect(component).not.toBeNull();
    });

    test('Set data submenu to Menu', () => {
        const TEST_DATA1 = 'test1';
        const TEST_DATA2 = 'test2';
        const TEST_DATA3 = 'test3';
        const mockData = [
            {
                id: TEST_DATA1,
                name: TEST_DATA1,
                title: TEST_DATA1,
                data: [
                    {
                        id: TEST_DATA2,
                        name: TEST_DATA2,
                        'data-testid': TEST_DATA2,
                    },
                    {
                        id: TEST_DATA3,
                        name: TEST_DATA3,
                        'data-testid': TEST_DATA3,
                    },
                ],
            },
        ];
        const wrapper = render(<Menu data={mockData} />);
        const component1 = screen.getByTitle(TEST_DATA1);
        const component2 = wrapper.getByTestId(TEST_DATA2);
        const component3 = wrapper.getByTestId(TEST_DATA3);

        expect(component1).not.toBeNull();
        expect(component2).not.toBeNull();
        expect(component3).not.toBeNull();
    });

    test('Set children with click event', () => {
        const TEST_FN = jest.fn();
        const TEST_JSX = <div data-testid={TEST_ID}>test</div>;
        const wrapper = render(<Menu onClick={TEST_FN}>{TEST_JSX}</Menu>);
        const jsx = wrapper.getByTestId(TEST_ID);

        fireEvent.click(jsx);
        expect(jsx).not.toBeNull();
        expect(TEST_FN).toBeCalled();
    });
    //
    test('Global Click EventListener', async () => {
        const TEST_DATA1 = 'test1';
        const TEST_DATA2 = 'test2';
        const TEST_DATA3 = 'test3';
        const mockData = [
            {
                id: TEST_DATA1,
                name: TEST_DATA1,
                title: TEST_DATA1,
                data: [
                    {
                        id: TEST_DATA2,
                        name: TEST_DATA2,
                        'data-testid': TEST_DATA2,
                    },
                    {
                        id: TEST_DATA3,
                        name: TEST_DATA3,
                        'data-testid': TEST_DATA3,
                    },
                ],
            },
        ];
        let component1: HTMLElement = document.createElement('li');

        Object.defineProperty(document, 'elementsFromPoint', {
            value: () => {
                return [component1];
            },
            writable: true,
        });
        render(<Menu trigger="click" data={mockData} />);
        component1 = screen.getByTitle(TEST_DATA1);

        fireEvent.click(component1);

        await waitFor(async () => {
            const component = screen.getByTitle(TEST_DATA1);
            expect(component.classList).toContain(activeClassName);

            fireEvent.click(window);
            await waitFor(() => {
                expect(component.classList).not.toContain(activeClassName);
            });
        });
    });

    test('Global ContextMenu EventListener', async () => {
        const TEST_DATA1 = 'test1';
        const TEST_DATA2 = 'test2';
        const TEST_DATA3 = 'test3';
        const mockData = [
            {
                id: TEST_DATA1,
                name: TEST_DATA1,
                title: TEST_DATA1,
                data: [
                    {
                        id: TEST_DATA2,
                        name: TEST_DATA2,
                        'data-testid': TEST_DATA2,
                    },
                    {
                        id: TEST_DATA3,
                        name: TEST_DATA3,
                        'data-testid': TEST_DATA3,
                    },
                ],
            },
        ];
        let component1: HTMLElement = document.createElement('li');

        Object.defineProperty(document, 'elementsFromPoint', {
            value: () => {
                return [component1];
            },
            writable: true,
        });
        render(<Menu trigger="click" data={mockData} />);
        component1 = screen.getByTitle(TEST_DATA1);

        fireEvent.click(component1);

        await waitFor(async () => {
            const component = screen.getByTitle(TEST_DATA1);
            expect(component.classList).toContain(activeClassName);

            fireEvent.contextMenu(window);
            await waitFor(() => {
                expect(component.classList).not.toContain(activeClassName);
            });
        });
    });

    test('Dispose the Menu', () => {
        const TEST_DATA1 = 'test1';
        const TEST_DATA2 = 'test2';
        const mockData = [
            {
                id: TEST_DATA1,
                name: TEST_DATA1,
                title: TEST_DATA1,
                data: [
                    {
                        id: TEST_DATA2,
                        name: TEST_DATA2,
                        'data-testid': TEST_DATA2,
                    },
                ],
            },
        ];
        const menu = renderer.create(<MenuTest />);
        const menuNode: any = (
            menu as renderer.ReactTestRenderer
        ).root.findByType(Menu);
        expect(menuNode._fiber).not.toBeUndefined();

        const menuRef = menuNode._fiber.ref;
        render(<Menu trigger="click" ref={menuRef} data={mockData} />);
        expect(menuRef?.current?.dispose).not.toBeUndefined();

        menuRef.current?.dispose();
        const item = document.body.querySelectorAll('ul')[1];
        expect(item.style.opacity).toEqual('0');
    });
});
