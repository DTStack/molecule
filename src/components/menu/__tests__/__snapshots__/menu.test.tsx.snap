// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the Menu Component Match the List snapshot 1`] = `
<ul
  className="mo-menu mo-menu--vertical"
  onClick={[Function]}
  onContextMenu={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  style={
    Object {
      "width": 200,
    }
  }
>
  <li
    className="mo-menu__item"
    data-mode="vertical"
    data-submenu={true}
    id="File"
  >
    <div
      className="mo-menu__content"
    >
      <span
        className="mo-menu__label"
      >
        File
      </span>
      <span
        className="mo-menu__indicator codicon codicon-chevron-right"
      />
    </div>
    <ul
      className="mo-menu mo-menu--vertical mo-sub-menu"
      id="File"
      style={
        Object {
          "opacity": "0",
          "pointerEvents": "none",
        }
      }
    >
      <li
        className="mo-menu__item"
        id="New File"
        onClick={[Function]}
      >
        <div
          className="mo-menu__content"
        >
          <span
            className="mo-menu__label"
            title="New File"
          >
            New File
          </span>
        </div>
      </li>
      <li
        className="mo-menu__item"
        id="OpenFile"
        onClick={[Function]}
      >
        <div
          className="mo-menu__content"
        >
          <span
            className="mo-menu__label"
            title="Open"
          >
            Open
          </span>
        </div>
      </li>
    </ul>
  </li>
  <li
    className="mo-menu__item"
    data-mode="vertical"
    data-submenu={true}
    id="Edit"
  >
    <div
      className="mo-menu__content"
    >
      <span
        className="mo-menu__label"
      >
        Edit
      </span>
      <span
        className="mo-menu__indicator codicon codicon-chevron-right"
      />
    </div>
    <ul
      className="mo-menu mo-menu--vertical mo-sub-menu"
      id="Edit"
      style={
        Object {
          "opacity": "0",
          "pointerEvents": "none",
        }
      }
    >
      <li
        className="mo-menu__item"
        id="Undo"
        onClick={[Function]}
      >
        <div
          className="mo-menu__content"
        >
          <span
            className="mo-menu__label"
            title="Undo"
          >
            Undo
          </span>
        </div>
      </li>
      <li
        className="mo-menu__item"
        id="Redo"
        onClick={[Function]}
      >
        <div
          className="mo-menu__content"
        >
          <span
            className="mo-menu__label"
            title="Redo"
          >
            Redo
          </span>
        </div>
      </li>
    </ul>
  </li>
  <li
    className="mo-menu__item"
    data-mode="vertical"
    data-submenu={true}
    id="Selection"
  >
    <div
      className="mo-menu__content"
    >
      <span
        className="mo-menu__label"
      >
        Selection
      </span>
      <span
        className="mo-menu__indicator codicon codicon-chevron-right"
      />
    </div>
    <ul
      className="mo-menu mo-menu--vertical mo-sub-menu"
      id="Selection"
      style={
        Object {
          "opacity": "0",
          "pointerEvents": "none",
        }
      }
    >
      <li
        className="mo-menu__item"
        id="SelectAll"
        onClick={[Function]}
      >
        <div
          className="mo-menu__content"
        >
          <span
            className="mo-menu__label"
            title="Select All"
          >
            Select All
          </span>
        </div>
      </li>
      <li
        className="mo-menu__item"
        id="CopyLineUp"
        onClick={[Function]}
      >
        <div
          className="mo-menu__content"
        >
          <span
            className="mo-menu__label"
            title="Copy Line Up"
          >
            Copy Line Up
          </span>
        </div>
      </li>
    </ul>
  </li>
  <li
    className="mo-menu__item"
    data-mode="vertical"
    data-submenu={true}
    id="View"
  >
    <div
      className="mo-menu__content"
    >
      <span
        className="mo-menu__label"
      >
        View
      </span>
      <span
        className="mo-menu__indicator codicon codicon-chevron-right"
      />
    </div>
    <ul
      className="mo-menu mo-menu--vertical mo-sub-menu"
      id="View"
      style={
        Object {
          "opacity": "0",
          "pointerEvents": "none",
        }
      }
    >
      <li
        className="mo-menu__item"
        id="Command Palette"
        onClick={[Function]}
      >
        <div
          className="mo-menu__content"
        >
          <span
            className="mo-menu__label"
            title="Command Palette"
          >
            Command Palette
          </span>
        </div>
      </li>
      <li
        className="mo-menu__item"
        id="OpenView"
        onClick={[Function]}
      >
        <div
          className="mo-menu__content"
        >
          <span
            className="mo-menu__label"
            title="Open View"
          >
            Open View
          </span>
        </div>
      </li>
      <li
        className="mo-menu__item"
        data-mode="vertical"
        data-submenu={true}
        id="Appearance"
      >
        <div
          className="mo-menu__content"
        >
          <span
            className="mo-menu__label"
          >
            Appearance
          </span>
          <span
            className="mo-menu__indicator codicon codicon-chevron-right"
          />
        </div>
        <ul
          className="mo-menu mo-menu--vertical mo-sub-menu"
          id="Appearance"
          style={
            Object {
              "opacity": "0",
              "pointerEvents": "none",
            }
          }
        >
          <li
            className="mo-menu__item"
            id="ShowMenuBar"
            onClick={[Function]}
          >
            <div
              className="mo-menu__content"
            >
              <span
                className="mo-menu__check codicon codicon-check"
              />
              <span
                className="mo-menu__label"
                title="Show Menu Bar"
              >
                Show Menu Bar
              </span>
            </div>
          </li>
          <li
            className="mo-menu__item"
            id="ShowSideBar"
            onClick={[Function]}
          >
            <div
              className="mo-menu__content"
            >
              <span
                className="mo-menu__check codicon codicon-check"
              />
              <span
                className="mo-menu__label"
                title="Show Side Bar"
              >
                Show Side Bar
              </span>
            </div>
          </li>
          <li
            className="mo-menu__item"
            id="ShowStatusBar"
            onClick={[Function]}
          >
            <div
              className="mo-menu__content"
            >
              <span
                className="mo-menu__check codicon codicon-check"
              />
              <span
                className="mo-menu__label"
                title="Show Status Bar"
              >
                Show Status Bar
              </span>
            </div>
          </li>
          <li
            className="mo-menu__item"
            id="ShowActivityBar"
            onClick={[Function]}
          >
            <div
              className="mo-menu__content"
            >
              <span
                className="mo-menu__check codicon codicon-check"
              />
              <span
                className="mo-menu__label"
                title="Show Activity Bar"
              >
                Show Activity Bar
              </span>
            </div>
          </li>
        </ul>
      </li>
    </ul>
  </li>
  <li
    className="mo-menu__item"
    data-mode="vertical"
    data-submenu={true}
    id="Run"
  >
    <div
      className="mo-menu__content"
    >
      <span
        className="mo-menu__label"
      >
        Run
      </span>
      <span
        className="mo-menu__indicator codicon codicon-chevron-right"
      />
    </div>
    <ul
      className="mo-menu mo-menu--vertical mo-sub-menu"
      id="Run"
      style={
        Object {
          "opacity": "0",
          "pointerEvents": "none",
        }
      }
    >
      <li
        className="mo-menu__item"
        id="RunTask"
        onClick={[Function]}
      >
        <div
          className="mo-menu__content"
        >
          <span
            className="mo-menu__label"
            title="Run Task"
          >
            Run Task
          </span>
        </div>
      </li>
    </ul>
  </li>
  <li
    className="mo-menu__item"
    data-mode="vertical"
    data-submenu={true}
    id="Help"
  >
    <div
      className="mo-menu__content"
    >
      <span
        className="mo-menu__label"
      >
        Help
      </span>
      <span
        className="mo-menu__indicator codicon codicon-chevron-right"
      />
    </div>
    <ul
      className="mo-menu mo-menu--vertical mo-sub-menu"
      id="Help"
      style={
        Object {
          "opacity": "0",
          "pointerEvents": "none",
        }
      }
    >
      <li
        className="mo-menu__item"
        id="About"
        onClick={[Function]}
      >
        <div
          className="mo-menu__content"
        >
          <span
            className="mo-menu__label"
            title="About"
          >
            About
          </span>
        </div>
      </li>
    </ul>
  </li>
</ul>
`;

exports[`Test the Menu Component Match the divider snapshot 1`] = `
<li
  className="mo-menu__item mo-menu__item--disabled"
  role="separator"
>
  <a
    className="mo-menu__divider"
  />
</li>
`;
